src/MemoryTracing/MemoryTracingIntegrationTest.cpp

```diff
@@ -0,0 +1,129 @@
+  }
+
+  [[nodiscard]] std::vector<ProducerCaptureEvent> GetAndClearEvents() {
+    absl::MutexLock lock{&events_mutex_};
+    std::vector<ProducerCaptureEvent> events = std::move(events_);
+    events_.clear();
+    return events;
+  }
+
+ private:
+  std::vector<ProducerCaptureEvent> events_;
+  absl::Mutex events_mutex_;
+};
+
+class MemoryTracingIntegrationTestFixture {
+ public:
+  MemoryTracingIntegrationTestFixture(uint64_t memory_sampling_period_ns)
+      : memory_sampling_period_ns_(memory_sampling_period_ns) {}
+
+  void StartTracingAndWaitForTracingLoopStarted() {
+    CHECK(!producer_.has_value());
+    CHECK(!listener_.has_value());
+    producer_.emplace(memory_sampling_period_ns_);
+    listener_.emplace();
+    producer_->SetListener(&*listener_);
+    producer_->Start();
+  }
+
+  [[nodiscard]] std::vector<ProducerCaptureEvent> StopTracingAndGetEvents() {
+    CHECK(producer_.has_value());
+    CHECK(listener_.has_value());
+    producer_->Stop();
+    producer_.reset();
+    std::vector<ProducerCaptureEvent> events = listener_->GetAndClearEvents();
+    listener_.reset();
+    return events;
+  }
+
+ private:
+  uint64_t memory_sampling_period_ns_;
+  std::optional<MemoryInfoProducer> producer_ = std::nullopt;
+  std::optional<BufferMemoryInfoListener> listener_ = std::nullopt;
+};
+
+[[nodiscard]] std::vector<ProducerCaptureEvent> TraceAndGetEvents(
+    MemoryTracingIntegrationTestFixture* fixture) {
+  CHECK(fixture != nullptr);
+
+  fixture->StartTracingAndWaitForTracingLoopStarted();
+
+  constexpr absl::Duration kSleepBeforeStopTracing = absl::Nanoseconds(1000000000);
+  absl::SleepFor(kSleepBeforeStopTracing);
+  return fixture->StopTracingAndGetEvents();
+}
+
+void VerifyOrderAndSamplingPeriodOfEvents(const std::vector<ProducerCaptureEvent>& events,
+                                          uint64_t sampling_period_ns) {
+  constexpr uint64_t kSamplingTimeErrorToleranceNs = 1000000;
+  uint64_t previous_event_timestamp_ns = 0;
+  for (const auto& event : events) {
+    if (event.event_case() != ProducerCaptureEvent::kSystemMemoryUsage) UNREACHABLE();
+
+    uint64_t current_event_timestamp_ns = event.system_memory_usage().timestamp_ns();
+    EXPECT_GE(current_event_timestamp_ns, previous_event_timestamp_ns);
+
+    if (previous_event_timestamp_ns > 0 && sampling_period_ns > 0) {
+      uint64_t time_since_last_successful_sampling =
+          current_event_timestamp_ns - previous_event_timestamp_ns;
+      uint64_t periods_since_last_successful_sampling =
+          time_since_last_successful_sampling / sampling_period_ns;
+      // If there are failed attempts of memory sampling, the sampling time error would be
+      // difficult to decide. Hence, we only check the case of successive successful attempts of
+      // sampling.
+      if (periods_since_last_successful_sampling <= 1) {
+        EXPECT_GE(time_since_last_successful_sampling,
+                  sampling_period_ns - kSamplingTimeErrorToleranceNs);
```

Comment:
this kind of seems flaky - I would say check number of events received against expected with +-1/2. It should be much less flaky since the main loop in the sampling thread is timed by deadlines.

Reply:
It is also difficult to decide the error tolerance of number of events, i.e., `+- n`, since we allow the failure of extracting `SystemMemoryUsage` from `/proc/meminfo`. When the `memory_sampling_period_ns_` becomes small, `# of received events - # of expected received events` would increase.

Reply:
> since we allow the failure of extracting SystemMemoryUsage from /proc/meminfo

Is this something we normally expect? I would imagine not, unless something is super strange about the machine, I guess `/proc/meminfo` should always be well formed.

Reply:
I think checking for count is somewhat less flaky because the implementation compensates for wait time flactuations. And we expect it to.

Reply:
Thanks for the explanation! Changed to check the received event counts.